### DISCUSSION TITLE
Provide a specific Mapper for mapping of property

### DISCUSSION THREAD

**Discussion by ghost** [riok/mapperly#1648] [Q&A] [answered]
I have an Object with 4 Lists of navigational properties.
I need to clone this object and while doing so reset all IDs of all items in the list to 0
I have written 4 mappers that look like this (just posting anonymized Entity / Method names):

    [MapValue(nameof(Entity1.ID), default)]
    public partial Entity1 CopyEntity1 (Entity1 model);

    [MapValue(nameof(Entity2.ID), default)]
    public partial Entity2 CopyEntity2 (Entity2 model);

    [MapValue(nameof(Entity3.ID), default)]
    public partial Entity3 CopyEntity3 (Entity3 model);

    [MapValue(nameof(Entity4.ID), default)]
    public partial Entity4 CopyEntity4 (Entity4 model);

However, for businesslogic reasons in other parts I have one more for one of the 4 Lists that does something else with the same input and needs to be the default since it is used in many different parts, which looks like this:

    [UserMapping(Default = true)]
    public partial Entity4 MapEntity4 (Entity4 model);

Since this is the default, when I generate the Code for the Clone Mapper of the parent Entity it results in the following auto-generated Code:

            target.Entity1.EnsureCapacity(model.Entity1.Count + target.Entity1.Count);
            foreach (var item in model.Entity1)
            {
                target.Entity1.Add(CopyEntity1(item));
            }

            target.Entity2.EnsureCapacity(model.Entity2.Count + target.Entity2.Count);
            foreach (var item in model.Entity2)
            {
                target.Entity2.Add(CopyEntity2(item));
            }            

            target.Entity3.EnsureCapacity(model.Entity3.Count + target.Entity3.Count);
            foreach (var item in model.Entity3)
            {
                target.Entity3.Add(CopyEntity3(item));
            }

            target.Entity4.EnsureCapacity(model.Entity4.Count + target.Entity4.Count);
            foreach (var item4 in model.Entity4)
            {
                target.Entity4.Add(MapEntity4(item4));
            }
            return target;

How can I tell the mapper to use CopyEntity4 instead of MapEntity4?

I've tried    
 [MapPropertyFromSource(nameof(Entity.Entity4), Use = nameof(CopyEntity4))]
 [MapProperty(nameof(Entity.Entity4), nameof(Entity.Entity4), Use = nameof(CopyEntity4))]

but both result in different errors.
Any help is appreciated

**Comment by latonz** [comment]
The types of the `MapProperty.Use` referenced method have to match exactly the types of the referenced properties. Therefore you would need an additional custom list mapping:
```csharp
[MapProperty(nameof(Entity.Entity4), nameof(Entity.Entity4), Use = nameof(CopyEntity4List))]
...

private List<Entity4> CopyEntity4List(List<Entity4> l) => l.Select(x => CopyEntity4(x)).ToList(); 
```

**Reply by ghost** [reply]
That seems like a valid approach, unfortunately it does not work.
I don't get an error, it compiles and builds, but it still uses MapEntity4 in the generated code

**Reply by latonz** [reply]
Can you post your declaration and generated code?

**Reply by ghost** [reply]
To expand on that, I researched the problem further, it seems to be because the list is readonly.
Your solution works as soon as it no longer readonly, then the generated code reads:

        target.Entity4 = CopyEntity4List(model.Entity4);

The List should stay readonly however, so what I would like to achieve is either:

        foreach (var item4 in model.Entity4)
        {
            target.Entity4.Add(CopyEntity4(item4));
        }

or 

        target.Entity4.AddRange(CopyEntity4List(model.Entity4));

**Reply by ghost** [reply]
This is the full clone method of the parent object (without the MapProperty for now since it didn't work)

    [UserMapping(Default = true)]
    [MapperRequiredMapping(RequiredMappingStrategy.Target)]
    [MapValue(nameof(Entity.ID), default)]
    [MapValue(nameof(Entity.ParentID), default)]
    [MapValue(nameof(Entity.TemplateID), default)]
    [MapValue(nameof(Entity.UpdateVon), Use = nameof(MapCurrentUser))]
    [MapValue(nameof(Entity.UpdateAm), Use = nameof(MapCurrentDateTime))]
    [MapValue(nameof(Entity.ErstelltVon), Use = nameof(MapCurrentUser))]
    [MapValue(nameof(Entity.ErstelltAm), Use = nameof(MapCurrentDateTime))]
    [MapperIgnoreSource(nameof(Entity.Parent))]
    public partial Entity Clone(Entity model);
    
This is the copy method of the sub entity

    [MapperRequiredMapping(RequiredMappingStrategy.Target)]
    [MapValue(nameof(Entity4.ID), default)]
    [MapperIgnoreTarget(nameof(Entity4.NavProp1))]
    [MapperIgnoreTarget(nameof(Entity4.NavProp2))]
    [MapperIgnoreTarget(nameof(Entity4.NavProp3))]
    public partial Entity4 CopyEntity4(Entity4 model);

this is the actual code, just with anonymized names of Entites and Properties due to our company guidelines
Also since it might be relevant, the mapper itself uses

    [Mapper(UseDeepCloning = true)]

**Reply by latonz** [reply, marked as answer]
Ah, I see. To achieve this, you would need support for existing target mappings combined with MapProperty.Use, which Mapperly currently does not support. A potential workaround is to separate the mappers by creating a dedicated Copy*Mapper class. This way, you can isolate these mappings and avoid any collisions.

I created https://github.com/riok/mapperly/issues/1649.

**Reply by ghost** [reply]
Okay, I actually thought about this already, but thought I might just have missed something.
Thanks you for clarifying!
Support for this would certainly be nice, but for now I can create a dedicated mapper.
